### PR TITLE
[TOOLS-1896] Touch Time by Issues

### PR DESCRIPTION
**Pull request by marcokuchla**
Touch Time by Issues Back-end and Front-end;

### CODE REVIEW

File: src/main/java/objective/taskboard/controller/TouchTimeKPIController.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            final String message = String.format("Project not found: %s.", projectKey);
+            return new ResponseEntity<>(message, HttpStatus.NOT_FOUND);
+        }
+        
+        ZoneId timezone = determineTimeZoneId(zoneId);
+        
+        final KpiLevel kpiLevel;
+        try {
+            kpiLevel = KpiLevel.valueOf(level.toUpperCase());
+        } catch (IllegalArgumentException e) {
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/controller/TouchTimeKPIController.java]: MAJOR Either log or rethrow this exception. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import objective.taskboard.followup.kpi.properties.KPIProperties;
+import objective.taskboard.jira.properties.JiraProperties;
+
+@Service
+public class TouchTimeKPIDataProvider {
+
+    private static final long HOUR_IN_SECONDS = 1 * 60 * 60;
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java]: MINOR Cast one of the operands of this multiplication operation to a "long". [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2184)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/objective/taskboard/controller/TouchTimeKPIController.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            final String message = String.format("Project not found: %s.", projectKey);
+            return new ResponseEntity<>(message, HttpStatus.NOT_FOUND);
+        }
+        
+        ZoneId timezone = determineTimeZoneId(zoneId);
+        
+        final KpiLevel kpiLevel;
+        try {
+            kpiLevel = KpiLevel.valueOf(level.toUpperCase());
+        } catch (IllegalArgumentException e) { //NOSONAR
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/controller/TouchTimeKPIController.java]: MAJOR Either log or rethrow this exception. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import objective.taskboard.followup.kpi.properties.KPIProperties;
+import objective.taskboard.jira.properties.JiraProperties;
+
+@Service
+public class TouchTimeKPIDataProvider {
+
+    private static final int HOUR_IN_SECONDS = 1 * 60 * 60;
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java]: MINOR Cast one of the operands of this multiplication operation to a "long". [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2184)

File: src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+                        .withChartType('bar')
+                        .withTooltipNumberOfDecimals(2)
+                        .withCategories(this.plotData.issueKeys)
+                        .withSeriesData(this.plotData.series);
+                    this.chart = builder.build();
+                },
+
+                _parseData: function (data) {
+                    data.forEach((dataPoint) => {
+                        dataPoint.effortInHours = Number(dataPoint.effortInHours);
```
Comment (jhonypg) [inline src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html]: Is it necessary?

File: src/test/java/objective/taskboard/controller/TouchTimeKPIControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    
+    private String projectKey;
+
+    private String zoneId;
+
+    private String level;
+    
+    @Before
+    public void setup() {
+        projectKey = "TEST";
```
Comment (jhonypg) [inline src/test/java/objective/taskboard/controller/TouchTimeKPIControllerTest.java]: Transform this setup into a DSL.

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    
+    @Mock
+    private IssueKpiService issueKpiService;
+
+    @InjectMocks
+    private TouchTimeKPIDataProvider subject = new TouchTimeKPIDataProvider();
+
+    @Before
+    public void setup() {
+        String[] progressingStatuses = new String[] { "UATing","Doing","QAing","Feature Reviewing","Alpha Testing","Reviewing" };
```
Comment (jhonypg) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java]: Transform this setup into a DSL.

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    private KPIProperties kpiProperties;
+    
+    @Mock
+    private JiraProperties jiraProperties;
+    
+    @Mock
+    private IssueKpiService issueKpiService;
+
+    @InjectMocks
+    private TouchTimeKPIDataProvider subject = new TouchTimeKPIDataProvider();
```
Comment (jhonypg) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java]: ` = new TouchTimeKPIDataProvider()`

Isn't necessary.

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        
+        Map<KpiLevel, List<IssueKpi>> allIssues = new HashMap<>();
+        allIssues.put(KpiLevel.SUBTASKS, issues);
+        
+        Mockito.when(issueKpiService.getIssuesFromCurrentState(projectKey, ZONE_ID)).thenReturn(allIssues);
+
+        TouchTimeChartDataSet actual = subject.getDataSet(projectKey, issueLevel, ZONE_ID);
+        assertThat(actual.points.size(), is(0));
+    }
+
```
Comment (jhonypg) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java]: Create a test mocking issue with KpiLevel that has not to show.
Something like "getTouchTimeChartDataSet_whenIssueIsNotFromRequestedLevel_thenRemoveItFromTheDataSet"

File: src/main/java/objective/taskboard/controller/TouchTimeKPIController.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
+package objective.taskboard.controller;
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply src/main/java/objective/taskboard/controller/TouchTimeKPIController.java]: Done

File: src/main/java/objective/taskboard/controller/TouchTimeKPIController.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    private ProjectDashboardTacticalPermission dashboardTacticalPermission;
+
+    @Autowired
+    private ProjectDashboardOperationalPermission projectDashboardOperationalPermission;
+    
+    @Autowired
+    private ProjectService projectService;
+    
+    @RequestMapping(value = "/api/projects/{project}/followup/touchtime/byissues", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Object> data(
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply src/main/java/objective/taskboard/controller/TouchTimeKPIController.java]: Done

File: src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcokuchla) [reply src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java]: Done

File: src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    private JiraProperties jiraProperties;
+
+    @Autowired
+    private KPIProperties kpiProperties;
+
+    public TouchTimeChartDataSet getDataSet(String projectKey, KpiLevel kpiLevel, ZoneId timezone) {
+        final Map<KpiLevel, List<IssueKpi>> allIssues = issueKpiService.getIssuesFromCurrentState(projectKey, timezone);
+        final List<String> statuses = getStatusesFromLevel(kpiLevel);
+        final List<IssueKpi> issues = allIssues.get(kpiLevel);
+        List<TouchTimeDataPoint> points = transformToDataPoints(issues, statuses);
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply src/main/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProvider.java]: Done

File: src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html (hunk 5 above)
Comment (marcokuchla) [reply src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html]: No. So it was removed.

File: src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html [hunk 13]
```diff
@@ -0,0 +1,263 @@
+<dom-module id="widget-touch-time">
+
+    <template>
+
+        <style>
+            .widget__body {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html]: Removed

File: src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+                    return Object.freeze({
+                        issueTypes: issueTypes,
+                        issueKeys: issueKeys,
+                        series: series,
+                        typeDimension: typeDimension
+                    });
+                },
+
+                createChart: function () {
+                    const builder = new TouchTimeChartBuilder('touchTime')
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply src/main/resources/static/elements/taskboard/kpis/widget-touch-time.html]: Suggest accepted

File: src/test/java/objective/taskboard/controller/TouchTimeKPIControllerTest.java (hunk 6 above)
Comment (marcokuchla) [reply src/test/java/objective/taskboard/controller/TouchTimeKPIControllerTest.java]: Done

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java (hunk 7 above)
Comment (marcokuchla) [reply src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java]: Done

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java (hunk 8 above)
Comment (marcokuchla) [reply src/test/java/objective/taskboard/followup/kpi/TouchTimeKPIDataProviderTest.java]: Removed

File: src/main/java/objective/taskboard/controller/TouchTimeKPIController.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+            @PathVariable("project") String projectKey,
+            @RequestParam("timezone") String zoneId,
+            @RequestParam("level") String level) {
+        
+        if (!dashboardTacticalPermission.isAuthorizedFor(projectKey)
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply src/main/java/objective/taskboard/controller/TouchTimeKPIController.java]: Done

**Pull request merged**